controllers/topology/desired_state.go

```diff
@@ -136,15 +138,86 @@ func computeControlPlane(class *clusterTopologyClass, current *clusterTopologySt
 		}
 	}
 
+	// Computes the desired Kubernetes version for the control plane.
+	controlPlaneVersion, err := computeControlPlaneVersion(current)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute spec.version for the ControlPlane object")
+	}
+
 	// Sets the desired Kubernetes version for the control plane.
-	// TODO: improve this logic by adding support for version upgrade component by component
-	if err := unstructured.SetNestedField(controlPlane.UnstructuredContent(), current.cluster.Spec.Topology.Version, "spec", "version"); err != nil {
+	if err := unstructured.SetNestedField(controlPlane.UnstructuredContent(), controlPlaneVersion, "spec", "version"); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion computes the ControlPlane version based on the current clusterTopologyState.
+// TODO: we also have to handle the following cases:
+// * ControlPlane.spec.version != MachineDeployment[].spec.template.spec.version, i.e. one of the MachineDeployments has
+//   a different version then the ControlPlane.
+func computeControlPlaneVersion(current *clusterTopologyState) (string, error) {
+	topologyVersion := current.cluster.Spec.Topology.Version
+
+	// ControlPlane does not exist yet, use topologyVersion for new ControlPlane.
+	if current.controlPlane.object == nil {
+		return topologyVersion, nil
+	}
+
+	controlPlaneVersion, ok, err := getControlPlaneVersion(current.controlPlane.object, "spec", "version")
+	if err != nil {
+		return "", err
+	}
+	if !ok {
+		return "", errors.Errorf("failed to get .spec.version from ControlPlane: not found")
+	}
+	controlPlaneStatusVersion, statusVersionSet, err := getControlPlaneVersion(current.controlPlane.object, "status", "version")
+	if err != nil {
+		return "", err
+	}
+
+	// Do not trigger another rollout, while a rollout is already in progress. A rollout is still in progress if:
+	// * .status.version is not set
+	// * .status.version is set but not equal to .spec.version
+	if !statusVersionSet || controlPlaneVersion != controlPlaneStatusVersion {
+		return controlPlaneVersion, nil
+	}
```

Comment:
This is obviously based on the assumption that we don't want to trigger another rollout while one is in progress. 
I'm not sure if that's what we want.